draft-ietf-masque-connect-udp-listen.md

```diff
@@ -143,30 +141,64 @@ this is the target host to which the proxy will send this UDP payload. When sent
 from proxy to client, this represents the source IP address of the UDP packet
 received by the proxy. This field has a length of 32 bits when the corresponding
 IP Version field value is 4, and 128 when the IP Version is 6.
-MAY be omitted if the latest previously transmitted Target IP Address and
-Port for the Context ID are the same as the current target IP Address and Port.
-If Omitted, the UDP Port and IP Version fields MUST also be omitted.
+It MUST be omitted if Context ID is non zero.
 
 UDP Port:
 
 : The UDP Port of this proxied UDP packet in network byte order. When sent from
 client to proxy, this is the target port to which the proxy will send this UDP
 payload. When sent from proxy to client, this represents the source UDP port of
-the UDP packet received by the proxy. MAY be omitted if the latest previously
-transmitted Target IP Address and Port for the Context ID are the same as the
-current target IP Address and Port. If Omitted, the IP Address and IP Version
-fields MUST also be omitted.
+the UDP packet received by the proxy. It MUST be omitted if Context ID is non zero.
 
 UDP Payload:
 
 : The unmodified UDP Payload of this proxied UDP packet (referred to as "data
 octets" in {{UDP}}).
 
 
+# Context ID and Header Compression {#contextid}
+The context ID is a 62 bit, variable length integer and MUST be specified in each UDP Payload carrying datagram. Context ID of value 0 represents no compression. The client must specify the IP version, IP and port of the target per datagram, see {{format}}.
+
+If the client intends to compress the IP and port information of the target, it must first register a new Context ID with the proxy, and specify a target IP and port via a Compression request Capsule, and the proxy confirms by responding with a Compression Assign CAPSULE see {{capsules}}. Hereon, both the client and the server MAY use that Context ID in its datagrams to suppress all IP/Port information to or from said target respectively.
+
+The IP Length, Address and Port fields are the same as those defined in {{format}}.
+
+# Capsules {#capsules}
+This document defines two new capsule types to enable registering target compression information, see {{capsulerequestformat}}:
+
+~~~ ascii-art
+Capsule {
+  Type COMPRESSION_REQUEST (0x04),
+  Length (i),
+  Target Mapping,
+}
+~~~
+{: #capsulerequestformat title="Compression Request Capsule Format"}
+
+~~~ ascii-art
+Capsule {
+  Type COMPRESSION_ASSIGN (0x05),
+  Length (i),
```

Comment:
Since we have COMPRESSION_ASSIGN, does it make sense to have COMPRESSION_REJECT? This is because a compression context requires the receiver to allocate memory, so to defend against memory exhaustion attacks it might be useful to say "nope sorry I'm not going to allocate these resources"